[Doc] V1 user guide (#13991)
vllm-project/vllm: docs/source/getting_started/v1_user_guide.md

#### Deprecated Features

As part of the major architectural rework in vLLM V1, several legacy features have been deprecated.

**Sampling features**

- **best_of**: This feature has been deprecated due to limited usage. See details at [RFC #13361](https://github.com/vllm-project/vllm/issues/13361).
- **Per-Request Logits Processors**: In V0, users could pass custom
  processing functions to adjust logits on a per-request basis. In vLLM V1 this feature has been deprecated. Instead, the design is moving toward supporting **global logits
  processors**, a feature the team is actively working on for future releases. See details at [RFC #13360](https://github.com/vllm-project/vllm/pull/13360).

**KV Cache features**

- **GPU <> CPU KV Cache Swapping**: with the new simplified core architecture, vLLM V1 no longer requires KV cache swapping

> to handle request preemptions.

Reviewer: To confirm: Have we decided to deprecate (i.e. no longer support) this feature? The reason I'm asking is there are some RFCs about adding CPU kv cache offloading to v1. cc @simon-mo @WoosukKwon

Author: Swapping will be deprecated, but offloading will not. 
cc @ywang96

> ### Feature & Model Support in Progress

Reviewer: One suggestion to this section is that it might be more clear to have an overview supported matrix: features vs status, where status could be one of the following:

- Deprecated: No plan to support in v1 unless there's a strong motivation.
- Planned: Plan to support but haven't started workings
- WIP: Being supported.
- Functional (unoptimized): It's working but is being optimized.
- Optimized: It's almost optimized and no other planned work atm.

So that in the rest of this section we could be feature centric to describe the current status of each feature and point to the corresponding GitHub issue/PR/Project.

Author: will do

Although the support for many features & models have been re‐implemented or optimized compared to V0, some remain either unsupported or not yet fully optimized on vLLM V1.

> #### Unoptimized Features

Reviewer: Better to overview that optimized features are the features that have been supported in v1 but we are still working in progress to optimize it.

- **LoRA**: LoRA is functionally working on vLLM V1 but its performance is
  inferior to that of V0. The team is actively working on improving its
  performance
(e.g., see [PR #13096](https://github.com/vllm-project/vllm/pull/13096)).

- **Spec Decode**: Currently, only ngram-based spec decode is supported in V1. There

>   will be follow-up work to support other types of spec decode.

Reviewer: Could link to the GitHub Project of v1 spec decode

#### Unsupported Features

- **FP8 KV Cache**: While vLLM V1 introduces new FP8 kernels for model weight quantization, support for an FP8 key–value cache is not yet available. Users must continue using FP16 (or other supported precisions) for the KV cache.

> - **Structured Generation Fallback**: For structured output tasks, V1 currently
>   supports only the `xgrammar:no_fallback` mode.
>   Details about the structured generation can be found
>   [here](https://docs.vllm.ai/en/latest/features/structured_outputs.html).

Reviewer: ```suggestion
- **Structured Output**: For structured output tasks, V1 currently
  supports only the `xgrammar:no_fallback` mode, meaning that it will error out if the output schema is unsupported by xgrammar.
  Details about the structured generation can be found
  [here](https://docs.vllm.ai/en/latest/features/structured_outputs.html).
```

#### Unsupported Models

> vLLM V1 excludes model architectures with the `SupportsV0Only` protocol, and the majority fall into the following categories. V1 support for these models will be added eventually.

Reviewer: ```suggestion
vLLM V1 currently excludes model architectures with the `SupportsV0Only` protocol, and the majority fall into the following categories. V1 support for these models will be added eventually.
```

# vLLM V1 User Guide

## Why vLLM V1?

vLLM V0 successfully supported a wide range of models and hardware, but as new features were developed independently, the system grew increasingly complex. This complexity made it harder to integrate new capabilities and introduced technical debt, revealing the need for a more streamlined and unified design.

Building on V0’s success, vLLM V1 retains the stable and proven components from V0
(such as the models, GPU kernels, and utilities). At the same time, it significantly
re-architects the core systems, covering the scheduler, KV cache manager, worker,
sampler, and API server, to provide a cohesive, maintainable framework that better
accommodates continued growth and innovation.

Specifically, V1 aims to:

- Provide a **simple, modular, and easy-to-hack codebase**.
- Ensure **high performance** with near-zero CPU overhead.
- **Combine key optimizations** into a unified architecture.
- Require **zero configs** by enabling features/optimizations by default.

For more details, check out the vLLM V1 blog post [vLLM V1: A Major
Upgrade to vLLM’s Core Architecture](https://blog.vllm.ai/2025/01/27/v1-alpha-release.html) (published Jan 27, 2025).

> This living user guide outlines a few known **important changes and limitations** introduced by vLLM V1. The team has been working actively to bring V1 as the default engine version of vLLM, therefore this guide will be updated constantly as more feature get supported on vLLM V1.

Reviewer 2: ```suggestion
This living user guide outlines a few known **important changes and limitations** introduced by vLLM V1. The team has been working actively to bring V1 as the default engine, therefore this guide will be updated constantly as more features get supported on vLLM V1.
```

### Semantic Changes and Deprecated Features

#### Logprobs

vLLM V1 supports logprobs and prompt logprobs. However, there are some important semantic
differences compared to V0:

**Logprobs Calculation**

Logprobs in V1 are now returned immediately once computed from the model’s raw output (i.e.
before applying any logits post-processing such as temperature scaling or penalty
adjustments). As a result, the returned logprobs do not reflect the final adjusted
probabilities used during sampling.

> Support for logprobs with post-sampling adjustments is work in progress and will be added in future updates.

Reviewer 2: ```suggestion
Support for logprobs with post-sampling adjustments is in progress and will be added in future updates.
```

**Prompt Logprobs with Prefix Caching**

Currently prompt logprobs are only supported when prefix caching is turned off via `--no-enable-prefix-caching`. In a future release, prompt logprobs will be compatible with prefix caching, but a recomputation will be triggered to recover the full prompt logprobs even upon a prefix cache hit. See details in [RFC #13414](https://github.com/vllm-project/vllm/issues/13414).

#### Deprecated Features

As part of the major architectural rework in vLLM V1, several legacy features have been deprecated.

**Sampling features**

- **best_of**: This feature has been deprecated due to limited usage. See details at [RFC #13361](https://github.com/vllm-project/vllm/issues/13361).
- **Per-Request Logits Processors**: In V0, users could pass custom

>   processing functions to adjust logits on a per-request basis. In vLLM V1 this feature has been deprecated. Instead, the design is moving toward supporting **global logits

Reviewer 2: ```suggestion
  processing functions to adjust logits on a per-request basis. In vLLM V1, this feature has been deprecated. Instead, the design is moving toward supporting **global logits
```

  processors**, a feature the team is actively working on for future releases. See details at [RFC #13360](https://github.com/vllm-project/vllm/pull/13360).

**KV Cache features**

- **GPU <> CPU KV Cache Swapping**: with the new simplified core architecture, vLLM V1 no longer requires KV cache swapping
to handle request preemptions.

### Feature & Model Support in Progress

Although we have re-implemented and partially optimized many features and models from V0 in vLLM V1, optimization work is still ongoing for some, and others remain unsupported.

#### Features to be Optimized

These features are already supported in vLLM V1, but their optimization is still
in progress.

- **LoRA**: LoRA is functionally working on vLLM V1 but its performance is
  inferior to that of V0. The team is actively working on improving its
  performance
(e.g., see [PR #13096](https://github.com/vllm-project/vllm/pull/13096)).

- **Spec Decode**: Currently, only ngram-based spec decode is supported in V1. There

>   will be follow-up work to support other types of spec decode (e.g., see [PR #13933](https://github.com/vllm-project/vllm/pull/13933)).

Reviewer 2: ```suggestion
  will be follow-up work to support other types of spec decode (e.g., see [PR #13933](https://github.com/vllm-project/vllm/pull/13933)). We will prioritize the support for Eagle, MTP compared to draft model based spec decode.
```

# vLLM V1 User Guide

> 

Reviewer 3: Add a note in the front saying the following
"V1 is enabled by default for supported use cases! We will gradually enable V1 for all use cases we plan to support. Please leave us feedback in GitHub or in vLLM Slack"

Building on V0’s success, vLLM V1 retains the stable and proven components from V0
(such as the models, GPU kernels, and utilities). At the same time, it significantly
re-architects the core systems, covering the scheduler, KV cache manager, worker,
sampler, and API server, to provide a cohesive, maintainable framework that better
accommodates continued growth and innovation.

Specifically, V1 aims to:

- Provide a **simple, modular, and easy-to-hack codebase**.
- Ensure **high performance** with near-zero CPU overhead.
- **Combine key optimizations** into a unified architecture.
- Require **zero configs** by enabling features/optimizations by default.

For more details, check out the vLLM V1 blog post [vLLM V1: A Major
Upgrade to vLLM’s Core Architecture](https://blog.vllm.ai/2025/01/27/v1-alpha-release.html) (published Jan 27, 2025).

This living user guide outlines a few known **important changes and limitations** introduced by vLLM V1. The team has been working actively to bring V1 as the default engine, therefore this guide will be updated constantly as more features get supported on vLLM V1.

> 

Reviewer 3: We need a paragraph on performance. Something like, we see significant performance improvements from upgrading to V1 core engine, in particular for long context scenarios. Please see performance benchmark [here](placeholder).

Author: added the placeholder for now.

### Feature / Model Supports Overview

> | Feature / Model                           | Status                                                                           |

Reviewer 3: Can we sort by status

| **Prompt Logprobs with Prefix Caching**   | <nobr>🟡 Planned ([RFC #13414](https://github.com/vllm-project/vllm/issues/13414))</nobr>  |
| **LoRA**                                  | <nobr>🟢 Functional ([PR #13096](https://github.com/vllm-project/vllm/pull/13096))</nobr>   |
| **Spec Decode**                           | <nobr>🚧 WIP ([PR #13933](https://github.com/vllm-project/vllm/pull/13933))</nobr>          |
| **FP8 KV Cache**                          | <nobr>🟡 Planned</nobr>                                                           |
| **Structured Generation Fallback**        | <nobr>🔴 Deprecated</nobr>                                                        |
| **best_of**                               | <nobr>🔴 Deprecated ([RFC #13361](https://github.com/vllm-project/vllm/issues/13361))</nobr>|
| **Per-Request Logits Processors**         | <nobr>🔴 Deprecated ([RFC #13360](https://github.com/vllm-project/vllm/pull/13360))</nobr>  |
| **GPU <> CPU KV Cache Swapping**          | <nobr>🔴 Deprecated</nobr>                                                        |

> | **Embedding Models**                      | <nobr>🟡 Planned</nobr>                                                           |

Reviewer 4: ```suggestion
| **Embedding Models**                      | <nobr>🟡 Planned <gh-issue:12249></nobr>                                                           |
```

- **Spec Decode**: Currently, only ngram-based spec decode is supported in V1. There
  will be follow-up work to support other types of spec decode (e.g., see [PR #13933](https://github.com/vllm-project/vllm/pull/13933)). We will prioritize the support for Eagle, MTP compared to draft model based spec decode.

#### Unsupported Features

- **FP8 KV Cache**: While vLLM V1 introduces new FP8 kernels for model weight quantization, support for an FP8 key–value cache is not yet available. Users must continue using FP16 (or other supported precisions) for the KV cache.

- **Structured Generation Fallback**: For structured output tasks, V1 currently
  supports only the `xgrammar:no_fallback` mode, meaning that it will error out if the output schema is unsupported by xgrammar.
  Details about the structured generation can be found
  [here](https://docs.vllm.ai/en/latest/features/structured_outputs.html).

#### Unsupported Models

vLLM V1 currently excludes model architectures with the `SupportsV0Only` protocol,
and the majority fall into the following categories. V1 support for these models will be added eventually.

**Embedding Models**  

> vLLM V1 does not yet include a `PoolingModelRunner` to support embedding/pooling
>   models (e.g, `XLMRobertaModel`).

Reviewer 4: ```suggestion
Instead of having a separate model runner, hidden states processor (<gh-issue:12249>), is proposed to enable simultaneous generation and pooling using the same engine.
```

# vLLM V1 User Guide

> 

Reviewer 5: A missing item is that fallback for structured generation is not supported. So only `xgrammar:no_fallback` is supported

Author: added in the unsupported features section

## Why vLLM V1?
Previous blog post [vLLM V1: A Major Upgrade to vLLM's Core Architecture](https://blog.vllm.ai/2025/01/27/V1-alpha-release.html)

## Semantic changes and deprecated features

### Logprobs
- vLLM V1 now supports both sample logprobs and prompt logprobs, as introduced in this [PR](https://github.com/vllm-project/vllm/pull/9880).
- **Current Limitations**: 
  - V1 prompt logprobs do not support prefix caching.
  - V1 logprobs are computed before logits post-processing, so penalty 
  adjustments and temperature scaling are not applied.
- The team is actively working on implementing logprobs that include post-sampling adjustments.

### The following features has been deprecated in V1:

#### Deprecated sampling features
- best_of

> - logits_processors

Reviewer 5: - "Per request logits processors are not supported"
- Nick is working on supporting these globally in the server

Author: does that means both best_of and logits_processors are both not deprecated but will be supported globally?

> - beam_search

Reviewer 5: I believe beam search is supported

#### Deprecated KV Cache features
- KV Cache swapping
- KV Cache offloading

## Unsupported features

- **LoRA**: LoRA works for V1 on the main branch, but its performance is inferior to that 
  of V0.
  The team is actively working on improving the performance [PR](https://github.com/vllm-project/vllm/pull/13096).

- **Spec Decode other than ngram**: currently, only ngram spec decode is supported in V1 
  after this [PR](https://github.com/vllm-project/vllm/pull/12193).

- **Quantization**: For V1, when the CUDA graph is enabled, it defaults to the 
  piecewise CUDA graph introduced in this [PR](https://github.com/vllm-project/vllm/pull/10058) ; consequently, FP8 and other quantizations are not supported. 

- **FP8 KV Cache**: FP8 KV Cache is not yet supported in V1.

> ## Unsupported models

Reviewer 5: - Mamba not supported yet
- We plan to support these eventually

# vLLM V1 User Guide

## Why vLLM V1?

> Previous blog post [vLLM V1: A Major Upgrade to vLLM's Core Architecture](https://blog.vllm.ai/2025/01/27/V1-alpha-release.html)

Reviewer 6: This link is invalid, showing "404", please check~

Reviewer 6: Did you mean: `https://blog.vllm.ai/2025/01/27/v1-alpha-release.html` ? please modify `V1` to `v1` in the link.

# vLLM V1 User Guide

## Why vLLM V1?

> Previous blog post [vLLM V1: A Major Upgrade to vLLM's Core Architecture](https://blog.vllm.ai/2025/01/27/v1-alpha-release.html)

Reviewer 7: Worth expanding here to give a bit more of the context and background - I will help you add stuff here.

> ## Semantic changes and deprecated features
> 

Reviewer 7: Maybe worth adding some context here (like how you did for `##Unsupported Models`. 

Can we also make sure the title format are consistent? 

```suggestion
## Semantic Changes and Deprecated Features
```

### Logprobs

> - vLLM V1 now supports both sample logprobs and prompt logprobs, as introduced in this [PR](https://github.com/vllm-project/vllm/pull/9880).
> - **Current Limitations**:

Reviewer 7: I think it's more natural to say something like

```suggestion
- vLLM V1 now supports both sample logprobs and prompt logprobs. Currently, the [implementation](https://github.com/vllm-project/vllm/pull/9880) has the following **limitations and semantic changes**:
```

  - V1 prompt logprobs do not support prefix caching.
  - V1 logprobs are computed before logits post-processing, so penalty
  adjustments and temperature scaling are not applied.
- The team is actively working on implementing logprobs that include post-sampling adjustments.

> ### The following features has been deprecated in V1:

Reviewer 7: ```suggestion
### Deprecated Features:
```

#### Current deprecated sampling features
The team is working on supporting these features globally in the server.

> 

Reviewer 7: This is not accurate since `best_of` and `per request logits processors` will be deprecated completely on V1.

Author: I saw @robertgshaw2-redhat 's comments on supporting them globally

Author: thank you @ywang96 I just realized I misunderstood Rob's comment.

- best_of
- per request logits processors

#### Deprecated KV Cache features
- KV Cache swapping
- KV Cache offloading

## Unsupported features

> 

Reviewer 7: Similar to my comment above for the first section: adding some context here will be helpful.

- **LoRA**: LoRA works for V1 on the main branch, but its performance is inferior to that
  of V0.
  The team is actively working on improving the performance [PR](https://github.com/vllm-project/vllm/pull/13096).

- **Spec Decode other than ngram**: currently, only ngram spec decode is supported in V1

>   after this [PR](https://github.com/vllm-project/vllm/pull/12193).

Reviewer 7: I don't think we really need to call out PR that's already merged.

- **Quantization**: For V1, when the CUDA graph is enabled, it defaults to the

>   piecewiseCUDA graphintroduced in this[PR](https://github.com/vllm-project/vllm/pull/10058); consequently,FP8 and other quantizations are not supported.

Reviewer 7: I don't think this is completely accurate? Also there are some issues in the formatting here.

- **FP8 KV Cache**: FP8 KV Cache is not yet supported in V1.

- **Structured Generation Fallback**: Only `xgrammar:no_fallback` is supported.
  Details about the structured generation can be found [here](https://docs.vllm.ai/en/latest/features/structured_outputs.html).

## Unsupported Models

vLLM V1 excludes models tagged with `SupportsV0Only` while we develop support for

> other types. For a complete list of supported models, see our [documentation](https://docs.vllm.ai/en/latest/models/supported_models.html). The

Reviewer 7: Let's remove `For a complete list of supported models, see our [documentation](https://docs.vllm.ai/en/latest/models/supported_models.html).` since you also have that at the end of the section.

following categories are currently unsupported, but we plan to
support them eventually.

**Embedding Models**
- vLLM V1 does not yet include a `PoolingModelRunner` to support embedding/pooling
  models.  
  *Example*: `BAAI/bge-m3`

**Mamba Models**  
- Models using selective state-space mechanisms (instead of standard transformer attention) are not yet supported.  
  *Examples*:  
    - Pure Mamba models (e.g., `BAAI/mamba-large`)  
    - Hybrid Mamba-Transformer models (e.g., `ibm-ai-platform/Bamba-9B`)

**Encoder-Decoder Models**  
- vLLM V1 is currently optimized for decoder-only transformers. Models requiring
  cross-attention between separate encoder and decoder (e.g.,

>   `facebook/bart-large-cnn`) are unsupported.

Reviewer 7: ```suggestion
  `facebook/bart-large-cnn`) are not yet supported.
```
